src/plugins/intel_cpu/src/ngraph_transformations/mha_fusion.cpp

```diff
@@ -0,0 +1,693 @@
+#include <ngraph/opsets/opset4.hpp>
+#include <ngraph/rt_info.hpp>
+#include <ngraph/pattern/op/wrap_type.hpp>
+#include "op/mha.hpp"
+
+#include "itt.hpp"
+
+// TODO: draw pattern
+ov::intel_cpu::MHAFloatFusion::MHAFloatFusion() {
+    MATCHER_SCOPE(MHAFloatFusion);
+
+    auto in0 = ngraph::pattern::any_input(ngraph::pattern::has_static_shape());
+    auto in1 = ngraph::pattern::any_input(ngraph::pattern::has_static_shape());
+    auto in2 = ngraph::pattern::wrap_type<ngraph::opset4::Constant>();
+    auto in3 = ngraph::pattern::any_input(ngraph::pattern::has_static_shape());
+    auto in4 = ngraph::pattern::wrap_type<ngraph::opset4::Constant>();
+    auto in5 = ngraph::pattern::wrap_type<ngraph::opset4::Constant>();
+    auto in6 = ngraph::pattern::wrap_type<ngraph::opset4::Constant>();
+    auto in7 = ngraph::pattern::wrap_type<ngraph::opset4::Constant>();
+    auto in8 = ngraph::pattern::any_input(ngraph::pattern::has_static_shape());
+    auto in9 = ngraph::pattern::wrap_type<ngraph::opset4::Constant>();
+    auto in10 = ngraph::pattern::wrap_type<ngraph::opset4::Constant>();
+    auto transpose0 = std::make_shared<ngraph::opset3::Transpose>(in0, in4);
+    auto transpose1 = std::make_shared<ngraph::opset3::Transpose>(in1, in5);
+    auto mul = std::make_shared<ngraph::opset3::Multiply>(transpose1, in2);
+    auto matmul0 = std::make_shared<ngraph::opset3::MatMul>(transpose0, mul);
+    auto add = std::make_shared<ngraph::opset4::Add>(matmul0, in3);
+    auto reshape0 = std::make_shared<ngraph::opset1::Reshape>(add, in6, true);
+    auto softmax = std::make_shared<ngraph::opset1::Softmax>(reshape0);
+    auto reshape1 = std::make_shared<ngraph::opset1::Reshape>(softmax, in7, true);
+    auto transpose2 = std::make_shared<ngraph::opset3::Transpose>(in8, in9);
+    auto matmul1 = std::make_shared<ngraph::opset3::MatMul>(reshape1, transpose2);
+    auto transpose3 = std::make_shared<ngraph::opset3::Transpose>(matmul1, in10);
+
+    ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher& m) {
+        auto& pattern_to_output = m.get_pattern_value_map();
+        auto transpose0_in = pattern_to_output.at(in0);
+        auto transpose1_in = pattern_to_output.at(in1);
+        auto mul_in1 = pattern_to_output.at(in2);
+        auto add_in1 = pattern_to_output.at(in3);
+        auto transpose2_in = pattern_to_output.at(in8);
+
+        if (transpose0_in.get_shape() != transpose1_in.get_shape() || transpose0_in.get_shape() != transpose2_in.get_shape()) {
+            return false;
+        }
+
+        if (transpose0_in.get_shape().size() != 4) {
+            return false;
+        }
+
+        auto expected_add_shape = Shape({transpose0_in.get_shape()[0], 1, 1, transpose0_in.get_shape()[1]});
+        if (add_in1.get_shape() != expected_add_shape) {
+            return false;
+        }
+
+        auto valid_transpose_order = [&](std::shared_ptr<Node> node, const std::vector<int64_t>& expected_order) {
```

Comment:
Could we create the common function for all matchers here to avoid code duplication? I see that we have the same lambda in each matcher

Reply:
Done: https://github.com/openvinotoolkit/openvino/pull/12936/commits/d3d717e58860dcac953d0408b5a327f04209f969